# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsChild.h
# Commit: c009bbdbbfd5
# Full Hash: c009bbdbbfd5451756cb1f3dc421b82df50911d1
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2019-11-05 22:07:34
# Regressor Bug: 1168606
# File Overlap Count: 1
# Description:
#   Bug 1168606 - Support preloading also for index cursors. r=ttung,asuth
#   
#   Depends on D43252
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D43461
# ==============================================================================

diff -r 89817a19f3b3 -r c009bbdbbfd5 dom/indexedDB/ActorsChild.h
--- a/dom/indexedDB/ActorsChild.h	Tue Nov 05 14:40:37 2019 +0000
+++ b/dom/indexedDB/ActorsChild.h	Tue Nov 05 14:40:35 2019 +0000
@@ -618,6 +618,9 @@
       const PreprocessParams& aParams) override;
 };
 
+// TODO: Consider defining different subclasses for the different cursor types,
+// possibly using the CRTP, which would remove the need for various case
+// distinctions.
 class BackgroundCursorChild final : public PBackgroundIDBCursorChild {
   friend class BackgroundTransactionChild;
   friend class BackgroundVersionChangeTransactionChild;
@@ -625,7 +628,11 @@
   class DelayedActionRunnable;
 
   struct CachedResponse {
-    CachedResponse() = default;
+    CachedResponse() = delete;
+
+    CachedResponse(Key aKey, StructuredCloneReadInfo&& aCloneInfo);
+    CachedResponse(Key aKey, Key aLocaleAwareKey, Key aObjectStoreKey,
+                   StructuredCloneReadInfo&& aCloneInfo);
 
     CachedResponse(CachedResponse&& aOther) = default;
     CachedResponse& operator=(CachedResponse&& aOther) = default;
@@ -633,7 +640,8 @@
     CachedResponse& operator=(const CachedResponse& aOther) = delete;
 
     Key mKey;
-    Key mObjectKey;  // TODO: This is not used anywhere right now
+    Key mLocaleAwareKey;
+    Key mObjectStoreKey;
     StructuredCloneReadInfo mCloneInfo;
   };
 
@@ -665,7 +673,8 @@
   }
 
   void SendContinueInternal(const CursorRequestParams& aParams,
-                            const Key& aCurrentKey);
+                            const Key& aCurrentKey,
+                            const Key& aCurrentObjectStoreKey);
 
   void SendDeleteMeInternal();
 
@@ -706,11 +715,18 @@
 
   void HandleResponse(const void_t& aResponse);
 
-  void HandleResponse(const nsTArray<ObjectStoreCursorResponse>& aResponse);
+  void HandleResponse(const nsTArray<ObjectStoreCursorResponse>& aResponses);
 
   void HandleResponse(const ObjectStoreKeyCursorResponse& aResponse);
 
-  void HandleResponse(const IndexCursorResponse& aResponse);
+  void HandleResponse(const nsTArray<IndexCursorResponse>& aResponses);
+
+  StructuredCloneReadInfo PrepareCloneReadInfo(
+      SerializedStructuredCloneReadInfo&& aCloneInfo) const;
+
+  template <typename T, typename Func>
+  void HandleMultipleCursorResponses(const nsTArray<T>& aResponses,
+                                     const Func& aHandleRecord);
 
   void HandleResponse(const IndexKeyCursorResponse& aResponse);
 
@@ -722,8 +738,8 @@
 
  public:
   // Force callers to use SendContinueInternal.
-  bool SendContinue(const CursorRequestParams& aParams,
-                    const Key& aCurrentKey) = delete;
+  bool SendContinue(const CursorRequestParams& aParams, const Key& aCurrentKey,
+                    const Key& aCurrentObjectStoreKey) = delete;
 
   bool SendDeleteMe() = delete;
 };